keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/VariantClasspathJar.kt

```diff
@@ -28,36 +34,100 @@ import kotlin.system.measureTimeMillis
+   */
+  @get:Internal
+  lateinit var configuration: Configuration
+
+  override fun copy() {
+    from(configuration.artifactView().files.filter { it.extension == "jar" }.map(project::zipTree))
+    super.copy()
+  }
+}
 
 /**
- * A [VariantClasspathJar] task that sources from both the androidTest compiled sources _and_ its distinct dependencies
- * (as compared to the [appRuntime]). R8's `PrintUses` requires no class overlap between the two jars it's comparing, so
+ * A [Jar] task that sources from both the androidTest compiled sources _and_ its distinct dependencies
+ * (as compared to the [appConfiguration]). R8's `PrintUses` requires no class overlap between the two jars it's comparing, so
  * at copy-time this will compute the unique androidTest dependencies. We need to have them because there may be
  * APIs that _they_ use that are used in the target app runtime, and we want R8 to account for those usages as well.
  */
-@Suppress("UnstableApiUsage")
 @CacheableTask
-abstract class AndroidTestVariantClasspathJar @Inject constructor(objects: ObjectFactory) : VariantClasspathJar() {
+abstract class AndroidTestVariantClasspathJar @Inject constructor(objects: ObjectFactory) : Jar() {
 
   private companion object {
     val LOG = AndroidTestVariantClasspathJar::class.simpleName!!
   }
 
+  /**
+   * This is the "official" input for wiring task dependencies correctly, but is otherwise
+   * unused. See [appConfiguration].
+   */
   @get:Classpath
-  val androidTestRuntime = objects.fileCollection()
+  val appArtifactFiles = objects.fileCollection()
 
+  /**
+   * This is what the task actually uses as its input.
+   */
+  @get:Internal
+  lateinit var appConfiguration: Configuration
+
+  /**
+   * This is the "official" input for wiring task dependencies correctly, but is otherwise
+   * unused. See [androidTestArtifactFiles].
+   */
   @get:Classpath
-  val appRuntime = objects.fileCollection()
+  val androidTestArtifactFiles = objects.fileCollection()
+
+  /** This is what the task actually uses as its input. */
+  @get:Internal
+  lateinit var androidTestConfiguration: Configuration
+
+  @get:Input
+  val emitDebugInfo: Property<Boolean> = objects.property()
 
   override fun copy() {
     measureTimeMillis {
       project.logger.debug("$LOG: Diffing androidTest jars and app jars")
-      val appJars = appRuntime.filter { it.extension == "jar" }.mapTo(mutableSetOf()) { it.nameWithoutExtension }
-      val distinctAndroidTestJars = androidTestRuntime.filter {
-        it.extension == "jar" && it.nameWithoutExtension !in appJars
+      val appJars = appConfiguration.artifactView().files.filterTo(LinkedHashSet()) { it.extension == "jar" }
+      val debug = emitDebugInfo.get()
+      if (debug) {
+        project.file("${project.buildDir}/${KeeperPlugin.INTERMEDIATES_DIR}/${archiveFile.get().asFile.nameWithoutExtension}AppJars.txt").apply {
+          writeText(appJars.sortedBy { it.path }
```

Comment:
Is it worth factoring out this block? The three usages in this function look pretty similar to me.

Reply:
c884e29